### PR TITLE
3ds: Add support for 3DS fingerprint fallback

### PR DESCRIPTION
**Pull request by jlejoux**
When 3DS2 native fingerprinting is not available for a specific transaction, fingerprinting needs to be done by loading a webpage. 

This adds the support of background fingerprinting by loading a hidden webview inside the app.
It adds the support of both `fingerprint` and `redirect` customer actions when performing an authorization request.

### CODE REVIEW

File: processout-sdk/src/main/java/com/processout/processout_sdk/ProcessOut.java [hunk 1]
```diff
@@ -233,20 +242,103 @@ public void error() {
...
+                            handler.onError(new ProcessOutException("Unsupported Android version"));
+                        }
+                        return super.shouldOverrideUrlLoading(view, request);
+                    }
+                });
+
+                // Start the timeout
+                timeOutHandler.postDelayed(timeoutClearer, 10000);
+
+                // Load the fingerprint URl
```
Comment (manuel-huez) [inline processout-sdk/src/main/java/com/processout/processout_sdk/ProcessOut.java]:
```suggestion
                // Load the fingerprint URL
```

File: processout-sdk/src/main/java/com/processout/processout_sdk/ProcessOut.java [hunk 2]
```diff
@@ -233,20 +242,103 @@ public void error() {
...
+                    public void run() {
+                        // Preparing the webview removal
+                        fingerPrintWebView.removeAllViews();
+                        fingerPrintWebView.clearHistory();
+                        fingerPrintWebView.clearCache(true);
+                        fingerPrintWebView.onPause();
+                        // Removing the webview and returning the error
+                        ((ViewGroup) fingerPrintWebView.getParent()).removeView(fingerPrintWebView);
+                        fingerPrintWebView.destroy();
+                        handler.onError(new ProcessOutNetworkException("Fingerprint fallback timed out."));
```
Comment (manuel-huez) [inline processout-sdk/src/main/java/com/processout/processout_sdk/ProcessOut.java]: You should rather fail with the appropriate token, so the user can continue the flow with a failed fingerprint

File: processout-sdk/src/main/java/com/processout/processout_sdk/ProcessOut.java [hunk 3]
```diff
@@ -233,20 +242,103 @@ public void error() {
...
+                                // The webview successfully received the gateway token
+                                // We cancel the timeout handler
+                                timeOutHandler.removeCallbacks(timeoutClearer);
+                                makeCardPayment(invoiceId, request.getUrl().getQueryParameter("token"), handler);
+                            }
+                        } else {
+                            // Android version not supported for fingerprinting
+                            // We cancel the timeout handler
+                            timeOutHandler.removeCallbacks(timeoutClearer);
+                            handler.onError(new ProcessOutException("Unsupported Android version"));
```
Comment (manuel-huez) [inline processout-sdk/src/main/java/com/processout/processout_sdk/ProcessOut.java]: You should rather fail with the appropriate token, so the user can continue the flow with a failed fingerprint

File: processout-sdk/src/main/java/com/processout/processout_sdk/ProcessOut.java [hunk 4]
```diff
@@ -233,20 +242,103 @@ public void error() {
...
+                        }
+                        return super.shouldOverrideUrlLoading(view, request);
```
Comment (manuel-huez) [inline processout-sdk/src/main/java/com/processout/processout_sdk/ProcessOut.java]: What does this do?

File: processout-sdk/src/main/java/com/processout/processout_sdk/ProcessOut.java [hunk 5]
```diff
@@ -233,20 +242,103 @@ public void error() {
...
+                break;
             default:
-                handler.onError(new ProcessOutAuthException("Unhandled three D S action:" + cA.getType().name()));
+                handler.onError(new ProcessOutException("Unhandled three D S action:" + cA.getType().name()));
                 break;
         }
     }
 
 
+    public void continueThreeDSVerification(final String invoiceId, String token, final ThreeDSVerificationCallback callback) {
```
Comment (manuel-huez) [inline processout-sdk/src/main/java/com/processout/processout_sdk/ProcessOut.java]: What does this function do? I can't see it used anywhere

Comment (jlejoux) [reply processout-sdk/src/main/java/com/processout/processout_sdk/ProcessOut.java]: It is used when coming back to the app here:
https://github.com/processout/processout-android/pull/8/files#diff-6b7d01fd02a31e2be5a308d336f14f91R47

File: processout-sdk/src/main/java/com/processout/processout_sdk/ProcessOut.java [hunk 6]
```diff
@@ -233,20 +242,108 @@ public void error() {
...
                 this.context.startActivity(browserIntent);
                 break;
+            case FINGERPRINT:
+                // Build the fingerprint hidden webview
+                final WebView fingerPrintWebView = new WebView(this.context);
+                fingerPrintWebView.getSettings().setJavaScriptEnabled(true); // enable javascript
+                fingerPrintWebView.setVisibility(View.INVISIBLE); // Hide the Webview
+
+                // Defining fallback request in case the fingerprint times out or is unavailable
+                MiscGatewayRequest fallbackGwayRequest = new MiscGatewayRequest("{\"threeDS2FingerprintTimeout\":true}");
```
Comment (manuel-huez) [inline processout-sdk/src/main/java/com/processout/processout_sdk/ProcessOut.java]: You also want to set its URL to the Fingerprint URL, as well as the headers to `Content-Type: application/json`

File: processout-sdk/src/main/java/com/processout/processout_sdk/ProcessOut.java (hunk 4 above)
Comment (jlejoux) [reply processout-sdk/src/main/java/com/processout/processout_sdk/ProcessOut.java]: This is required to specify if the webview should be loading the URL or not. In our case, the action has already been handled so we fallback to the default behavior.

**Pull request merged**